### PR TITLE
Multiprocessing added

### PR DESCRIPTION
**Pull request by rotelimelech**
* splitting the domain under poly_domains, and spawn a new enumerator for each chunk, under a different process
* each enumerator will only do the first enumeration. The refining is done in a single process
* to allow us to move actions from the second stage to the first (like calculating initial hits to a higher dept), I've added `multiprocessed_find_initial_hits` and `multiprocessed_refine_results` to the abstract enumerator
** in multiprocessed_find_initial_hits Im calling an "addtions" function, that allows child enumerators to add functionality to the first stage
** multiprocessed_refine_results completes this idea by discarding those actions form the second stage

### CODE REVIEW

**Comment by rotelimelech** [comment]
After your comments, i've changed the first and second step without adding abstract functions.
You're right, it looks much cleaner that way :)

File: ramanujan/poly_domains/CartesianProductPolyDomain.py [hunk 1]
```diff
@@ -89,10 +90,53 @@ def iter_polys(self, primary_looped_domain):
...
+		for i, v in enumerate(ranges):
+			ranges_metadata.append({
+				'range': v,
+				'size': v[1]-v[0]+1,
+				'series': series,
+				'index': i
+				})
+		return ranges_metadata
+
+	def split_domains_to_processes(self, number_of_instances):
```
Comment (ShaharGottlieb) [inline ramanujan/poly_domains/CartesianProductPolyDomain.py:113]: please add a unit test for this function, making sure that the split is done correctly - 
all elements exist in some partition, and  no more than the total number of elements expected

**Review by ShaharGottlieb — CHANGES_REQUESTED** [review]
overall nice job!
small fixes and i'll approve

File: ramanujan/multiprocess_enumeration.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    # Each subprocess only uses lhs.bloom. See Dummy class doc for more details.
+    lean_lhs = Dummy()
+    lean_lhs.bloom = lhs.bloom
+
+    # Creating arguments for each process function
+    split_domain = poly_search_domain.split_domains_to_processes(number_of_processes)
+    for domain_chunk in split_domain:
+        arguments.append((
+            enumerator_class,
+            deepcopy(lean_lhs.bloom),
```
[the remark this answers is not in the document]
Comment (rotelimelech) [reply ramanujan/multiprocess_enumeration.py]: it happens automatically, i've removed it

**Review by ShaharGottlieb — APPROVED** [review]

**Pull request merged**